### PR TITLE
Snippets and Auto-closing pairs play nicer

### LINKED ISSUE #104952: Snippets + Auto-closing pairs don't compliment each other
Version: 1.49.0-insider
Commit: 11dc5a81ba248cc2678888391f1b24dccabddaf8
Date: 2020-08-17T17:11:16.078Z (1 day ago)
Electron: 9.2.0
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Darwin x64 19.6.0

Steps to Reproduce:

Add the following python snippet:
```json
"listcompPy": {
		"prefix":"[x for x in",
		"body":[
			"[x for x in range(1,20) if x%2==0 ]"
		]
	}
```
or PowerShell snippet:
```json
"PSCustomObject": {
        "prefix":"[PSCustomObject",
        "body":[
            "[PSCustomObject] @{ Key = Value }"
        ]
    }
```

1. type: `[` which will cause auto-closing pair... (at this point you have `[|]` where `|` is the cursor)
2. then add a `x` or `psc` and hit TAB or ENTER on the intellisense item

#### Python

```
[x for x in range(1,20) if x%2==0 ]
```

#### PowerShell

```
[PSCustomObject]@{
    Name = Value
}
```

#### Python

```
[x for x in range(1,20) if x%2==0 ]]
```

#### PowerShell

```
[PSCustomObject]@{
    Name = Value
}]
```

NOTE THE TRAILING `]`. Snippets and auto-closing pairs should play nice together so that the whole experience generates valid code.

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
This PR fixes #104952

Previously, auto-closing pairs would get in the way of some snippets if the snippet's prefix started with an autoclosing pair (like `[`). Here's an example:

```json
"PSCustomObject": {
        "prefix":"[PSCustomObject",
        "body":[
            "[PSCustomObject] @{ Key = Value }"
        ]
    }
```

To fix/workaround this "odd behavior" (from an end user perspective) this PR uses the following logic to decide whether or not the snippet eats the character immidiately to the right of the cursor:

1. First check if there is anything to the right of the cursor
2. Then check If the character to the right of the cursor is a closing character of an autoclosing pair for this language
3. Then check if the start position in the file is the opening character of an autoclosing pair
4. Finally check if the snippet prefix contains the opening and closing pair at its edges

If this is satified, we say that the end of the snippet is endColumn++ thus eating the character to the right of the cursor.

This requires that the snippet more or less "opts-in" to this behavior by making sure that the prefix has autoclosing pairs around it... so if I wanted to opt-in the above snippet, I would change it to this:

```json
"PSCustomObject": {
        "prefix":"[PSCustomObject]",
        "body":[
            "[PSCustomObject] @{ Key = Value }"
        ]
    }
```

> Note the extra `]` in the prefix

Then, and only then will the eat the `]` of the auto-closing pair and _only_ if the user uses _replace_ for the suggestion (SHIFT+ENTER by default) instead of the _insert_ behavior of just ENTER. If the user wants ENTER to have this behavior, they can apply the following setting:

```json
"editor.suggest.insertMode": "replace"
```

which is made exactly for this purpose.

### CODE REVIEW

File: src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts [hunk 1]
```diff
@@ -43,11 +44,13 @@ suite('SnippetsService', function () {
 		});
 	});
 
+	let disposables: Array<IDisposable>;
```
Comment (jrieken) [inline src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts:47]: nit: We have a `DisposableStore` which makes this a little simpler

File: src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts [hunk 2]
```diff
@@ -68,6 +71,11 @@ suite('SnippetsService', function () {
 		)]);
 	});
 
+	teardown(function () {
+		for (const disposable of disposables) {
+			disposable.dispose();
+		}
```
Comment (jrieken) [inline src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts:77]: would then be `store.clear()`

**Review by jrieken — APPROVED** [review]

File: src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts (hunk 2 above)
Comment (TylerLeonhardt) [reply src/vs/workbench/contrib/snippets/test/browser/snippetsService.test.ts]: I refactored this a bit so there's only one DisposableStore for the whole test suite and then I dispose it in suiteTeardown

**Comment by ArturoDent** [comment]
In testing this snippet:

```jsonc
"PSCustomObject": {
        "prefix":"[PSCustomObject]",
        "body":[
            "[PSCustomObject] @{ Key = Value }"
        ]
 }
```

Here are my results with `suggest.insertMode` at the default `insert`:

	[PSCustomObject] @{ Key = Value }    // via tab, insertMode: default insert

	[PSCustomObject] @{ Key = Value }    // via enter, insertMode: default insert  <==== ??

	[PSCustomObject] @{ Key = Value }]   // via shift-enter, insertMode: default insert

So I am confused by the statement above:

**"Then, and only then will the eat the ] of the auto-closing pair and *only* if the user uses replace for the suggestion (SHIFT+ENTER by default) instead of the *insert* behavior of just ENTER."**

As you can see the end auto-closing pair was eaten for me using <kbd>Enter</kbd> and default `suggest.insertMode` of `insert`.  I like this behaviour but it seems to conflict with the quoted description??

I get the exact same results with `suggest.insertMode` set to `replace`:


     [PSCustomObject] @{ Key = Value }  // via tab, insertMode: replace

    [PSCustomObject] @{ Key = Value }  // via enter, insertMode: replace

    [PSCustomObject] @{ Key = Value }] // via shift-enter, insertMode: replace


Am I missing something or is the description inaccurate?

**Comment by TylerLeonhardt** [comment]
@ArturoDent are you sure your insertMode wasn't set to replace somewhere else? Because the default for Shift+Enter is replace....unless you set insertMode to replace in which case Shift+Enter does insert.

**Comment by ArturoDent** [comment]
`suggest.insertMode` is at the default `insert`:  


autoClosePairSnippet

--------------------

Version: 1.53.0-insider (user setup)
Commit: c265dff48a27868c7f0f84fd25a169897dc43d73
Date: 2021-01-14T06:31:42.082Z
Electron: 11.2.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.21286

**Comment by TylerLeonhardt** [comment]
Look at the suggestion tooltip in your gif. It says "ENTER" for Replace and SHIFT+ENTER for Insert.

Maybe you have a workspace setting?

**Comment by ArturoDent** [comment]
I have no workspace settings in that workspace.

**Comment by ArturoDent** [comment]
Here is a full list of results for a **JSON file**.  One set with both user and workspace `sugget.insertMode` set to `insert` and another with `suggest.insertMode` set to `replace`.

You see I tried both triggering the snippet with `[psc]` and `psc` - there was a different result only for <kbd>Shift</kbd>+<kbd>Enter</kbd> - should that be the case.

      "PSCustomObject": {
		"prefix": "[PSCustomObject]",
		"body": [
			"[PSCustomObject] @{ Key = Value }"
		]
	},

     [PSCustomObject] @{ Key = Value }     // tab, insertMode = insert, with [psc]
     [PSCustomObject] @{ Key = Value }     // tab, insertMode = replace, with [psc]

     [PSCustomObject] @{ Key = Value }     // tab, insertMode = insert, with psc
     [PSCustomObject] @{ Key = Value }     // tab, insertMode = replace, with psc

     [PSCustomObject] @{ Key = Value }     // enter, insertMode = insert, with [psc]
     [PSCustomObject] @{ Key = Value }     // enter, insertMode = replace, with [psc]

     [PSCustomObject] @{ Key = Value }     // enter, insertMode = insert, with psc
     [PSCustomObject] @{ Key = Value }     // enter, insertMode = replace, with psc

     [PSCustomObject] @{ Key = Value }]    // shift+enter, insertMode = insert, with [psc]
     [PSCustomObject] @{ Key = Value }]    // shift+enter, insertMode = replace, with [psc]

     [PSCustomObject] @{ Key = Value }     // shift+enter, insertMode = insert, with psc
     [PSCustomObject] @{ Key = Value }     // shift+enter, insertMode = replace, with psc

-------------------

Here are the results for a **non-JSON file**:

    [PSCustomObject] @{ Key = Value }]  // enter, insert [psc]
    [PSCustomObject] @{ Key = Value }   // enter, insert, psc

    [PSCustomObject] @{ Key = Value }   // shift+enter, insert [psc]
    [PSCustomObject] @{ Key = Value }   // shift+enter, insert psc

    [PSCustomObject] @{ Key = Value }]  // tab, insert [psc]
    [PSCustomObject] @{ Key = Value }   // tab, insert psc

    [PSCustomObject] @{ Key = Value }]  // enter, replace [psc]
    [PSCustomObject] @{ Key = Value }   // enter, replace psc

    [PSCustomObject] @{ Key = Value }]  // shift+enter, replace [psc]
    [PSCustomObject] @{ Key = Value }   // shift+enter, replace psc

    [PSCustomObject] @{ Key = Value }   // tab, replace [psc]
    [PSCustomObject] @{ Key = Value }   // tab, replace psc

------------------------

The second set of results is probably more in line with what you were expecting - although note there is no difference between `insert` and `replace` modes using the `psc` trigger whereas there are differences when using the `[psc]` trigger.

I know JSON completion is trickier because what is treated as a word is different in those files.

**Comment by TylerLeonhardt** [comment]
`Shift`+`Enter` and `Enter/Tab` should always have opposite behavior.

If `"editor.suggest.insertMode": "insert"`:

`Enter` should still show the `]` at the end
`Shift+Enter` should not

If `"editor.suggest.insertMode": "replace"`:

`Shift+Enter` should still show the `]` at the end
`Enter` should not

That is what is shown in the suggest status bar:
image

It seems like the problem here is that when you have `"editor.suggest.insertMode": "insert"` that behavior is not being honored. This even applies to other snippets without autoclosing pairs:

```
        "sdf": {
		"prefix": "asdf",
		"body": "ffffffff"
	}
```

I don't _think_ this PR caused this behavior but it's certainly supspicious.

**Comment by ArturoDent** [comment]
Using

        "sdf": {
		"prefix": "asdf",
		"body": "ffffffff"
	}

and `"editor.suggest.insertMode": "insert"`  

The only way I have found to actually `insert` and not always `replace` is the third example in the following demo.  Where I actually had to <kbd>Esc</kbd> to close the suggestion widget, reopen with <kbd>Ctrl</kbd>+<kbd>Space</kbd>  and then <kbd>Enter</kbd> or mouse-clicking the Insert hint does seem to work.  I say "seem" because while the `fffffff` is inserted at the cursor position the preceding part of the prefix, say `a` is replaced.  I am not sure whether that is expected behaviour?

autoClosePairSnippet2

**Comment by TylerLeonhardt** [comment]
> I am not sure whether that is expected behaviour?

Yes that all looks right. Keep in mind that "replace" behavior only decides _if the content to the right of the cursor is included in the prefix on the left of the cursor_. 

Everything to the left of the cursor is _always_ replaced.

**Pull request merged**